src/box/iproto.cc

```diff
@@ -2224,102 +2331,185 @@ struct iproto_cfg_msg: public cbus_call_msg
+{
+	cfg_msg->mem_used =
+		slab_cache_used(&iproto_thread->net_cord.slabc) +
+		slab_cache_used(&iproto_thread->net_slabc);
+	cfg_msg->connections =
+		mempool_count(&iproto_thread->iproto_connection_pool);
+	cfg_msg->requests =
+		mempool_count(&iproto_thread->iproto_msg_pool);
+}
+
 static int
 iproto_do_cfg_f(struct cbus_call_msg *m)
 {
 	struct iproto_cfg_msg *cfg_msg = (struct iproto_cfg_msg *) m;
 	int old;
+	struct iproto_thread *iproto_thread = cfg_msg->iproto_thread;
+
 	try {
 		switch (cfg_msg->op) {
 		case IPROTO_CFG_MSG_MAX:
-			cpipe_set_max_input(&tx_pipe,
+			cpipe_set_max_input(&iproto_thread->tx_pipe,
 					    cfg_msg->iproto_msg_max / 2);
 			old = iproto_msg_max;
 			iproto_msg_max = cfg_msg->iproto_msg_max;
 			if (old < iproto_msg_max)
-				iproto_resume();
+				iproto_resume(iproto_thread);
 			break;
 		case IPROTO_CFG_LISTEN:
-			if (evio_service_is_active(&binary))
-				evio_service_stop(&binary);
-			if (cfg_msg->uri != NULL &&
-			    (evio_service_bind(&binary, cfg_msg->uri) != 0 ||
-			     evio_service_listen(&binary) != 0))
+			if (evio_service_is_active(&iproto_thread->binary)) {
+				diag_set(ClientError, ER_UNSUPPORTED, "Iproto",
+					 "listen if service already active");
+				diag_raise();
+			}
+			iproto_thread_fill_binary(iproto_thread, cfg_msg->binary);
+			if (evio_service_listen(&iproto_thread->binary) != 0)
 				diag_raise();
-			cfg_msg->addrlen = binary.addr_len;
-			cfg_msg->addr = binary.addrstorage;
+			break;
+		case IPROTO_CFG_STOP:
+			if (evio_service_is_active(&iproto_thread->binary))
+				evio_service_stop(&iproto_thread->binary);
+			break;
+		case IPROTO_CFG_STAT:
+			iproto_fill_stat(iproto_thread, cfg_msg);
 			break;
 		default:
 			unreachable();
 		}
 	} catch (Exception *e) {
 		return -1;
 	}
+
 	return 0;
 }
 
 static inline void
-iproto_do_cfg(struct iproto_cfg_msg *msg)
+iproto_do_cfg(struct iproto_thread *iproto_thread, struct iproto_cfg_msg *msg)
 {
-	if (cbus_call(&net_pipe, &tx_pipe, msg, iproto_do_cfg_f,
-		      NULL, TIMEOUT_INFINITY) != 0)
+	msg->iproto_thread = iproto_thread;
+	if (cbus_call(&iproto_thread->net_pipe, &iproto_thread->tx_pipe, msg,
+		      iproto_do_cfg_f, NULL, TIMEOUT_INFINITY) != 0)
 		diag_raise();
 }
 
-void
-iproto_listen(const char *uri)
+static inline void
+iproto_stop(void)
+{
+	struct iproto_cfg_msg cfg_msg;
+	iproto_cfg_msg_create(&cfg_msg, IPROTO_CFG_STOP);
+	for (unsigned int i = 0; i < iproto_threads_count; i++)
+		iproto_do_cfg(&iproto_threads[i], &cfg_msg);
+}
+
+static inline void
+iproto_listen(struct evio_service *binary)
 {
 	struct iproto_cfg_msg cfg_msg;
 	iproto_cfg_msg_create(&cfg_msg, IPROTO_CFG_LISTEN);
-	cfg_msg.uri = uri;
-	iproto_do_cfg(&cfg_msg);
-	iproto_bound_address_storage = cfg_msg.addr;
-	iproto_bound_address_len = cfg_msg.addrlen;
+	cfg_msg.binary = binary;
+	for (unsigned int i = 0; i < iproto_threads_count; i++)
+		iproto_do_cfg(&iproto_threads[i], &cfg_msg);
+}
+
+void
+iproto_listen(const char *uri)
```

Comment:
Two funcs with the same name - I'd rather rename one of them

Reply:
Fix